scripts/data/client.py

```diff
@@ -125,101 +149,104 @@ def process_batch(job):
     else:
-        match = re.search(r"gas_spent=(\d+)", result.stdout)
+        match = re.search(r"gas_spent=(\d+)", stdout)
         gas_info = f"gas spent: {int(match.group(1))}" if match else "no gas info found"
         logger.info(f"{job} done, {gas_info}")
         if not match:
-            logger.warning(f"{job}: not gas info found")
+            logger.warning(f"{job}: no gas info found")
+
+
+def process_batch(job):
+    if shutdown_event.is_set():
+        logger.info(f"Shutdown detected, skipping job: {job}")
+        return
+
+    arguments_file = job.batch_file.as_posix().replace(".json", "-arguments.json")
+    try:
+        with open(arguments_file, "w") as af:
+            af.write(str(format_args(job.batch_file, False, False)))
+        run_process(arguments_file, job)
+    except Exception as e:
+        logger.error(f"Error processing batch {job}: {e}")
 
 
-# Producer function: Generates data and adds jobs to the queue
 def job_producer(job_gen):
     global current_weight
 
     try:
         for job, weight in job_gen:
-            # Wait until there is enough weight capacity to add the new block
+            if shutdown_event.is_set():
+                logger.info("Shutdown detected in producer. Stopping job production.")
+                break
+
             with weight_lock:
                 logger.debug(
                     f"Adding job: {job}, current total weight: {current_weight}..."
                 )
-                while (
-                    (current_weight + weight > MAX_WEIGHT_LIMIT)
-                    and current_weight != 0
+                while not shutdown_event.is_set() and (
+                    (current_weight + weight > MAX_WEIGHT_LIMIT and current_weight != 0)
                     or job_queue.full()
                 ):
                     logger.debug("Producer is waiting for weight to be released.")
-                    weight_lock.wait()  # Wait for the condition to be met
+                    weight_lock.wait(timeout=1)
 
-                if (current_weight + weight > MAX_WEIGHT_LIMIT) and current_weight == 0:
+                if shutdown_event.is_set():
+                    break
+
+                if current_weight + weight > MAX_WEIGHT_LIMIT and current_weight == 0:
                     logger.warning(f"{job} over the weight limit: {MAX_WEIGHT_LIMIT}")
 
-                # Add the job to the queue and update the weight
                 job_queue.put((job, weight))
                 current_weight += weight
-                logger.debug(
-                    f"Produced job: {job}, current total weight: {current_weight}"
-                )
-
-                # Notify consumers that a new job is available
                 weight_lock.notify_all()
+
     finally:
         logger.debug("Producer is exiting...")
-        for _ in range(THREAD_POOL_SIZE):
-            job_queue.put(None)
-
+        if not shutdown_event.is_set():
```

Comment:
Why is it necessary? If there are no more jobs we just let consumers finish, no?

Reply:
If you let this happen during shutdown you will not need to wait with timeout in line 224. Information that there is no more work will reach consumers.

Reply:
I've refactored Producer to now properly signal end of work to all consumers using None sentinels